src/gcp.rs

```diff
@@ -79,68 +82,90 @@ pub struct GcpAuthConfig {
+    None,
+}
 
 #[derive(Debug)]
-struct Inner {
-    creds: Option<Credentials>,
-    scope: Scope,
+pub struct InnerCreds {
+    creds: Option<(Credentials, Scope)>,
     token: RwLock<Token>,
 }
 
-impl GcpCredentials {
+impl GcpAuthenticator {
     async fn from_file(path: &str, scope: Scope) -> crate::Result<Self> {
         let creds = Credentials::from_file(path).context(InvalidCredentialsSnafu)?;
-        let token = fetch_token(&creds, &scope).await?;
-        Ok(Self(Arc::new(Inner {
-            creds: Some(creds),
-            scope,
-            token: RwLock::new(token),
-        })))
+        let token = RwLock::new(fetch_token(&creds, &scope).await?);
+        let creds = Some((creds, scope));
+        Ok(Self::Credentials(Arc::new(InnerCreds { creds, token })))
     }
 
-    async fn new_implicit(scope: Scope) -> crate::Result<Self> {
-        let token = get_token_implicit().await?;
-        Ok(Self(Arc::new(Inner {
-            creds: None,
-            scope,
-            token: RwLock::new(token),
-        })))
+    async fn new_implicit() -> crate::Result<Self> {
+        let token = RwLock::new(get_token_implicit().await?);
+        let creds = None;
+        Ok(Self::Credentials(Arc::new(InnerCreds { creds, token })))
     }
 
-    pub fn make_token(&self) -> String {
-        let token = self.0.token.read().unwrap();
-        format!("{} {}", token.token_type(), token.access_token())
+    fn from_api_key(api_key: &str) -> crate::Result<Self> {
+        base64::decode_config(api_key, base64::URL_SAFE).context(InvalidApiKeySnafu)?;
+        Ok(Self::ApiKey(api_key.into()))
+    }
+
+    pub fn make_token(&self) -> Option<String> {
+        match self {
+            Self::Credentials(inner) => Some(inner.make_token()),
+            Self::ApiKey(_) | Self::None => None,
+        }
     }
 
     pub fn apply<T>(&self, request: &mut http::Request<T>) {
-        request
-            .headers_mut()
-            .insert(AUTHORIZATION, self.make_token().parse().unwrap());
+        if let Some(token) = self.make_token() {
+            request
+                .headers_mut()
+                .insert(AUTHORIZATION, token.parse().unwrap());
+        }
+        self.apply_uri(request.uri_mut());
     }
 
-    async fn regenerate_token(&self) -> crate::Result<()> {
-        let token = match &self.0.creds {
-            Some(creds) => fetch_token(creds, &self.0.scope).await?,
-            None => get_token_implicit().await?,
-        };
-        *self.0.token.write().unwrap() = token;
-        Ok(())
+    pub fn apply_uri(&self, uri: &mut Uri) {
+        match self {
+            Self::Credentials(_) | Self::None => (),
+            Self::ApiKey(api_key) => {
+                let mut parts = uri.clone().into_parts();
+                let path = parts
+                    .path_and_query
+                    .as_ref()
+                    .map_or("/", PathAndQuery::path);
+                let paq = format!("{path}?key={api_key}");
```

Comment:
Wondering if we should percent encode the api_key just-in-case.

Reply:
The `from_api_key` function checks to ensure the `api_key` is valid base64, which is a subset of what can be parsed here, so it will never fail to parse nor require encoding.

Reply:
Ah, cool, thanks for confirming 👍